gateway/handle.go

```diff
@@ -614,3 +614,233 @@ func (gw *Handle) addToWebRequestQ(_ *http.ResponseWriter, req *http.Request, do
+					}
+					// skip calculate sdkVersion - expected to be done already
+				}
+				// skip bot agent check\
+			}
+
+			if isUserSuppressed(workspaceID, userIDFromReq, sourceID) {
+				gw.logger.Infow("suppressed event",
+					"userIDFromReq", userIDFromReq,
+					"sourceID", sourceID,
+					"workspaceID", workspaceID,
+				)
+				gw.eventFailed(arctx, reqType, errEventSuppressed.Error())
+				continue
+			}
+
+			rudderID, _ := misc.GetMD5UUID(userIDFromReq + ":" + anonIDFromReq)
+			toSet["rudderId"] = rudderID
+			if eventTypeFromReq == "audiencelist" {
+				containsAudienceList = true
+			}
+			userID := buildUserID(userIDHeader, anonIDFromReq, userIDFromReq)
+			out = append(out, jobObject{
+				userID: userID, events: []map[string]interface{}{toSet},
+			})
+		}
+		// some checks here
+		if gw.conf.enableRateLimit.Load() {
+			ok, errCheck := gw.rateLimiter.CheckLimitReached(ctx, workspaceID, int64(len(eventsBatch)))
+			if errCheck != nil {
+				gw.stats.NewTaggedStat(
+					"gateway.rate_limiter_error",
+					stats.CountType,
+					stats.Tags{"workspaceId": workspaceID},
+				).Increment()
+				gw.logger.Errorf("Rate limiter error: %v Allowing the request", errCheck)
+			}
+			if ok {
+				errorMessage = errRequestDropped.Error()
+				return
+			}
+		}
+		if len(out) == 0 { // events suppressed - but return success
+			return
+		}
+		// do we need this..?
+		if len(body) > gw.conf.maxReqSize.Load() && !containsAudienceList {
+			err = errors.New(response.RequestBodyTooLarge)
+			gw.requestFailed(arctx, reqType, err.Error())
+			return
+		}
+
+		params := map[string]any{
+			"source_id":          sourceID,
+			"source_job_run_id":  sourcesJobRunID,
+			"source_task_run_id": sourcesTaskRunID,
+		}
+		marshalledParams, err := json.Marshal(params)
+		if err != nil {
+			gw.logger.Errorf(
+				"[Gateway] Failed to marshal parameters map. Parameters: %+v",
+				params,
+			)
+			marshalledParams = []byte(
+				`{"error": "rudder-server gateway failed to marshal params"}`,
+			)
+		}
+		jobs := make([]*jobsdb.JobT, 0)
+		for _, userEvent := range out {
+			var (
+				payload    json.RawMessage
+				eventCount int
+			)
+			{
+				type SingularEventBatch struct {
+					Batch      []map[string]interface{} `json:"batch"`
+					RequestIP  string                   `json:"requestIP"`
+					WriteKey   string                   `json:"writeKey"`
+					ReceivedAt string                   `json:"receivedAt"`
+				}
+				receivedAt, ok := userEvent.events[0]["receivedAt"].(string)
+				if !ok || !arctx.ReplaySource {
+					receivedAt = time.Now().Format(misc.RFC3339Milli)
+				}
+				singularEventBatch := SingularEventBatch{
+					Batch:      userEvent.events,
+					RequestIP:  ipAddr, // TODO: processor gets IPAddr from inside each event
+					WriteKey:   arctx.WriteKey,
+					ReceivedAt: receivedAt,
+				}
+				payload, err = json.Marshal(singularEventBatch)
+				if err != nil {
+					panic(err)
+				}
+				eventCount = len(userEvent.events)
+			}
```

Comment:
Do we need an additional block in here?
```suggestion
				type SingularEventBatch struct {
					Batch      []map[string]interface{} `json:"batch"`
					RequestIP  string                   `json:"requestIP"`
					WriteKey   string                   `json:"writeKey"`
					ReceivedAt string                   `json:"receivedAt"`
				}
				receivedAt, ok := userEvent.events[0]["receivedAt"].(string)
				if !ok || !arctx.ReplaySource {
					receivedAt = time.Now().Format(misc.RFC3339Milli)
				}
				singularEventBatch := SingularEventBatch{
					Batch:      userEvent.events,
					RequestIP:  ipAddr, // TODO: processor gets IPAddr from inside each event
					WriteKey:   arctx.WriteKey,
					ReceivedAt: receivedAt,
				}
				payload, err = json.Marshal(singularEventBatch)
				if err != nil {
					panic(err)
				}
				eventCount = len(userEvent.events)
```